colexechash: switch to operating with uint32 values

yuzefovich opened the pull request:
This commit switches the hash table to operate on uint32 hash values
rather than uint64. This switch allows us to reduce the memory
footprint as well as to actually speed up the computation. Full
microbenchmarks are [here](https://gist.github.com/yuzefovich/8a672586343a903f2669fa883c67c265), but the gist is as follows:
- hash aggregator has speedup of 10-20% and reduced mem usage of 5-40%
- unordered distinct has speedup of 2-6% and reduced mem usage of 2-30%
- hash joiner has speedup of 2-25% and reduced mem usage of 4-10%

Difference on TPCH queries (average over 100 runs):
```
Q1:	before: 3.32s	after: 3.16s	 -4.93%
Q2:	before: 3.42s	after: 3.15s	 -8.08%
Q3:	before: 2.68s	after: 2.39s	 -10.88%
Q4:	before: 1.72s	after: 1.64s	 -4.46%
Q5:	before: 2.45s	after: 2.35s	 -4.22%
Q6:	before: 4.56s	after: 4.49s	 -1.49%
Q7:	before: 5.69s	after: 5.57s	 -2.16%
Q8:	before: 1.10s	after: 1.09s	 -0.66%
Q9:	before: 5.69s	after: 5.80s	 2.00%
Q10:	before: 2.09s	after: 1.98s	 -5.17%
Q11:	before: 0.95s	after: 0.91s	 -4.13%
Q12:	before: 4.26s	after: 4.28s	 0.57%
Q13:	before: 1.34s	after: 1.26s	 -6.20%
Q14:	before: 0.44s	after: 0.45s	 2.28%
Q15:	before: 2.47s	after: 2.47s	 -0.13%
Q16:	before: 0.93s	after: 0.88s	 -4.68%
Q17:	before: 0.25s	after: 0.24s	 -2.76%
Q18:	before: 2.18s	after: 2.12s	 -2.48%
Q19:	before: 0.51s	after: 0.49s	 -2.79%
Q20:	before: 9.80s	after: 9.53s	 -2.70%
Q21:	before: 4.96s	after: 4.85s	 -2.26%
Q22:	before: 0.58s	after: 0.57s	 -1.67%
```

The rationale for why this change is not breaking is that we don't expect
to need more than 4 billion buckets. For the hash table that would mean
that about 4 billion rows are buffered in memory which we should never
reach given the workmem limit that kicks in at around million of tuples
in the hash table.

Note that we still internally compute the hash of stuff using uint64,
but then later when we assign it to return outside of the hash functions
we trim it down to uint32.

However, in order to not change the routing behavior of the hash router
(which uses the tuple hash distributor that is also powered by the hash
utilities) we need to preserve uint64 usage in that case. We do so by
duplicating a couple of helper functions and making another function
generic. Note that we couldn't make `initHash` because we use
concretely-typed slice for fast zeroing out, but we could have made
`rehash` generic, however, that currently incurs a small performance
penalty, so we choose to duplicate the generated code there.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101579)
<!-- Reviewable:end -->

yuzefovich commented:
Ideally, I'd like to backport this to 23.1.0 given that we've done execinfra version bumps in 23.1 but will deprecate that versioning scheme in 23.2.

@cockroachdb/sql-queries I'd like to hear thoughts about that as well as to get feedback on the caveats that come with it:
> The only caveat this change has is that now it's possible for the hash
router to send rows to different outputs than was done before this
change (which could lead to incorrect query results). In other words,
we would need to gate this change based on some version. However,
during 23.1 release cycle we bumped the execinfra version a couple of
times, and I want to piggy-back on that. As a result of those bumps,
it's not possible to create a distributed plan in which some nodes are
running 22.2 while others are on 23.1.

> This would mean that clusters with mixed 23.1.0-beta.1 (which doesn't
have this change) and 23.1.0 (which might have this change) can produce
incorrect results for distributed queries. If backporting this change to
23.1.0 branch or this mixed-beta-version cluster incorrect results are
unacceptable, nothing is lost - we will simply not apply this change to
the tuple hash distributor and will just generate duplicated hash
functions.

yuzefovich commented:
> This would mean that clusters with mixed 23.1.0-beta.1 (which doesn't
have this change) and 23.1.0 (which might have this change) can produce
incorrect results for distributed queries.

This caveat can be addressed by bumping execinfra version in this PR, but then ideally we'd still backport it to 23.1.0 so that now new execinfra versions are introduced after 23.1 since we're deprecating this scheme.

rytaft commented:
Does this strictly need to be backported to 23.1.x at all? If it's not critical, I would suggest not backporting. If it is, sounds like 23.1.0 is significantly better than waiting to 23.1.1?





<!-- Sent from Reviewable.io -->

yuzefovich commented:
> Does this strictly need to be backported to 23.1.x at all?

No, we could keep the logic of the tuple hash distributor use uint64 meaning that the hash router's routing won't change. This will result in duplicating several hashing functions (generated into `colexechash/hash_utils.eg.go`).

I was hoping to backport this given that it brings impressive improvement on the microbenchmarks, but on TPCH queries the improvement seems minor (included in the PR description), so it seems ok to not backport this.

Hold off from reviewing for now.

yuzefovich committed:
colexechash: switch to operating with uint32 values

This commit switches the hash table to operate on uint32 hash values
rather than uint64. This switch allows us to reduce the memory
footprint as well as to actually speed up the computation. Full
microbenchmarks are [here](https://gist.github.com/yuzefovich/8a672586343a903f2669fa883c67c265),
but the gist is as follows:
- hash aggregator has speedup of 10-20% and reduced mem usage of 5-40%
- unordered distinct has speedup of 2-6% and reduced mem usage of 2-30%
- hash joiner has speedup of 2-25% and reduced mem usage of 4-10%

Difference on TPCH queries (average over 100 runs):
```
Q1:	before: 3.32s	after: 3.16s	 -4.93%
Q2:	before: 3.42s	after: 3.15s	 -8.08%
Q3:	before: 2.68s	after: 2.39s	 -10.88%
Q4:	before: 1.72s	after: 1.64s	 -4.46%
Q5:	before: 2.45s	after: 2.35s	 -4.22%
Q6:	before: 4.56s	after: 4.49s	 -1.49%
Q7:	before: 5.69s	after: 5.57s	 -2.16%
Q8:	before: 1.10s	after: 1.09s	 -0.66%
Q9:	before: 5.69s	after: 5.80s	 2.00%
Q10:	before: 2.09s	after: 1.98s	 -5.17%
Q11:	before: 0.95s	after: 0.91s	 -4.13%
Q12:	before: 4.26s	after: 4.28s	 0.57%
Q13:	before: 1.34s	after: 1.26s	 -6.20%
Q14:	before: 0.44s	after: 0.45s	 2.28%
Q15:	before: 2.47s	after: 2.47s	 -0.13%
Q16:	before: 0.93s	after: 0.88s	 -4.68%
Q17:	before: 0.25s	after: 0.24s	 -2.76%
Q18:	before: 2.18s	after: 2.12s	 -2.48%
Q19:	before: 0.51s	after: 0.49s	 -2.79%
Q20:	before: 9.80s	after: 9.53s	 -2.70%
Q21:	before: 4.96s	after: 4.85s	 -2.26%
Q22:	before: 0.58s	after: 0.57s	 -1.67%
```

The rationale for why this change is not breaking is that we don't expect
to need more than 4 billion buckets. For the hash table that would mean
that about 4 billion rows are buffered in memory which we should never
reach given the workmem limit that kicks in at around million of tuples
in the hash table.

Note that we still internally compute the hash of stuff using uint64,
but then later when we assign it to return outside of the hash functions
we trim it down to uint32.

However, in order to not change the routing behavior of the hash router
(which uses the tuple hash distributor that is also powered by the hash
utilities) we need to preserve uint64 usage in that case. We do so by
duplicating a couple of helper functions and making another function
generic. Note that we couldn't make `initHash` because we use
concretely-typed slice for fast zeroing out, but we could have made
`rehash` generic, however, that currently incurs a small performance
penalty, so we choose to duplicate the generated code there.

Epic: None

Release note: None

yuzefovich commented:
RFAL @DrewKimball @mgartner

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 20 files at r1, 14 of 14 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101579#-:-NV_zSHF9QCr1dgAf-lP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR Drew! And thanks Becca for the input earlier!

bors r+

yuzefovich commented:
bors r+
